fms_extras/utils/generation.py

```diff
@@ -3,9 +3,623 @@
+            the flattening operation if redundant tokens are being removed
+    """
+    if not flattening:
+        return input_ids.view(-1, input_ids.size(2)), None, None
+
+    # flatten the batch
+    flat_inputs, unflat_indices, flat_indices = flatten_batch(
+        input_ids
+    )  # n', b k 1+h, n'
+
+    # check if high enough level of compression to perform batch flattening
+    compression = flat_inputs.numel() / input_ids.numel()
+    perform_batch_flattening = compression < 0.75
+
+    # set the correct input_ids to return if compression ratio satisfied
+    if perform_batch_flattening:
+        result_input_ids = flat_inputs[None,]  # 1 n'
+
+        __apply_flattening_to_cache_metadata(cache_data, unflat_indices, flat_indices)
+    else:
+        result_input_ids = input_ids.view(-1, input_ids.size(2))
+        unflat_indices = None  # type: ignore
+        flat_indices = None  # type: ignore
+
+    return result_input_ids, unflat_indices, flat_indices
+
+
+def __free_incorrect_tokens_and_sequences(
+    kv_cache_manager: PagedKVCacheManager,
+    child_sequence_ids_list: List[List[int]],
+    best_guess: torch.Tensor,
+    n_correct: torch.Tensor,
+    decode_seq_length: int,
+) -> List[int]:
+    """
+    Free all non-best candidate sequences from the kv-cache and for best candidate sequences, logically remove tokens
+    based on the number of incorrect tokens generated
+
+    Args:
+        kv_cache_manager: PagedKVCacheManager
+            the paged attention kv-cache manager
+        child_sequence_ids_list: List[List[int]]
+            the list of child sequences per parent
+        best_guess: torch.Tensor
+            a tensor containing the best candidate per sequence in the batch
+        n_correct: torch.Tensor
+            a tensor containing the number of correct tokens best candidate in the batch
+        decode_seq_length: int
+            the model input length at decode step
+
+    Returns:
+    List[int]
+        a list of the new parent sequence ids for next decode step
+    """
+    parent_sequence_ids = []
+    for parent_index, child_sequence_ids in enumerate(child_sequence_ids_list):
+        best_index = int(best_guess[parent_index].item())
+
+        # free all bad candidates
+        kv_cache_manager.free_sequences(
+            child_sequence_ids[:best_index] + child_sequence_ids[best_index + 1 :]
+        )
+
+        # decrease the context length of the sequence which used to be sequence length + inp_len by the number of incorrect tokens
+        # for the best candidate
+        best_sequence_id = child_sequence_ids[best_index]
+        parent_sequence_ids.append(best_sequence_id)
+        kv_cache_manager.remove_tokens(
+            best_sequence_id, decode_seq_length - n_correct[parent_index].item() - 1  # type: ignore
+        )
+    return parent_sequence_ids
+
+
+def __apply_batch_expansion_to_cache_metadata(
```

Comment:
same here, might also make sense for it to be a method on cache_data

Reply:
fixed in latest commit